core/src/labels/labels.cpp

```diff
@@ -84,98 +69,218 @@ void Labels::update(const View& _view, float _dt, const std::vector<std::unique_
+
+                    for (auto& l1 : mesh1->getLabels()) {
+                        if (!l1 || !l1->canOcclude() || l0->hash() != l1->hash()) {
+                            continue;
+                        }
+                        float d2 = glm::distance2(l0->transform().state.screenPos,
+                                l1->transform().state.screenPos);
 
-            for (const auto& t1 : tiles) {
-                if (!t1) { continue; }
-                for (const auto& style : _styles) {
-                    const auto& m0 = t0->getMesh(*style);
-                    if (!m0) { continue; }
-                    const LabelMesh* mesh0 = dynamic_cast<const LabelMesh*>(m0.get());
-                    if (!mesh0) { continue; }
-                    const auto& m1 = t1->getMesh(*style);
-                    if (!m1) { continue; }
-                    const LabelMesh* mesh1 = static_cast<const LabelMesh*>(m1.get());
-
-                    for (auto& l0 : mesh0->getLabels()) {
-                        if (!l0->canOcclude()) { continue; }
-
-                        for (auto& l1 : mesh1->getLabels()) {
-                            if (!l1 || !l1->canOcclude() || l0->hash() != l1->hash()) {
-                                continue;
-                            }
-                            float d2 = glm::distance2(l0->transform().state.screenPos,
-                                    l1->transform().state.screenPos);
-
-                            // The new label lies within the circle defined by the bbox of l0
-                            if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
-                                l0->skipTransitions();
-                            }
+                        // The new label lies within the circle defined by the bbox of l0
+                        if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
+                            l0->skipTransitions();
                         }
                     }
                 }
             }
         }
     }
+}
+
+void Labels::checkRepeatGroups(std::vector<TextLabel*>& _visibleSet) const {
+    struct GroupElement {
+        glm::vec2 position;
+        float threshold;
+        const std::string* group;
+
+        bool operator==(const GroupElement& _ge) {
+            return _ge.position == position
+                && _ge.threshold == threshold
+                && *_ge.group == *group;
+        };
+    };
+
+    std::map<size_t, std::vector<GroupElement>> repeatGroups;
+
+    for (TextLabel* textLabel : _visibleSet) {
 
-    //// Update label meshes
+        size_t hash = std::hash<std::string>()(textLabel->options().repeatGroup);
+        GroupElement element;
+
+        element.position = textLabel->center();
+        element.threshold = textLabel->options().repeatDistance;
+        element.group = &textLabel->options().repeatGroup;
+
+        auto it = repeatGroups.find(hash);
+        if (it == repeatGroups.end()) {
```

Comment:
I'm not sure what you mean. This case adds the first item of a repeat group. (So there is no need to check if it conflicts with another labels of the same group)